gazebo_plugins/test/camera/triggered_camera.cpp

```diff
@@ -0,0 +1,113 @@
+#include <ros/ros.h>
+#include <std_msgs/Empty.h>
+
+class CameraTest : public testing::Test
+{
+protected:
+  virtual void SetUp()
+  {
+    images_received_ = 0;
+  }
+
+  ros::NodeHandle nh_;
+  image_transport::Subscriber cam_sub_;
+  int images_received_;
+  ros::Time image_stamp_;
+public:
+  void imageCallback(const sensor_msgs::ImageConstPtr& msg)
+  {
+    image_stamp_ = msg->header.stamp;
+    images_received_++;
+  }
+};
+
+// Test if the camera image is published at all, and that the timestamp
+// is not too long in the past.
+TEST_F(CameraTest, cameraSubscribeTest)
+{
+  image_transport::ImageTransport it(nh_);
+  cam_sub_ = it.subscribe("camera1/image_raw", 1,
+                          &CameraTest::imageCallback,
+                          dynamic_cast<CameraTest*>(this));
+
+  // wait for 3 seconds to confirm that we don't receive any images
+  for (unsigned int i = 0; i < 30; ++i)
+  {
+    ros::spinOnce();
+    ros::Duration(0.1).sleep();
+  }
+  EXPECT_EQ(images_received_, 0);
+
+  // make sure something is subscribing to image_trigger topic
+  // there is no easy API, so call getSystemState
+  XmlRpc::XmlRpcValue args, result, payload;
+  args[0] = ros::this_node::getName();
+  EXPECT_TRUE(ros::master::execute("getSystemState", args, result, payload, true));
+  // [publishers, subscribers, services]
+  // subscribers in index 1 of payload
+  int trigger_listeners = 0;
+  for (int i = 0; i < payload[1].size(); ++i)
+  {
+    // [ [topic1, [topic1Subscriber1...topic1SubscriberN]] ... ]
+    // topic name i is in index 0
+    std::string topic = payload[1][i][0];
+    if (topic.find("image_trigger") != std::string::npos)
+    {
+      trigger_listeners++;
+    }
+  }
+  EXPECT_EQ(trigger_listeners, 1);
+
+  // publish to trigger topic and expect an update within one second:
+  ros::Publisher trigger_pub = nh_.advertise<std_msgs::Empty>("camera1/image_trigger", 1, true);
+  std_msgs::Empty msg;
+
+  trigger_pub.publish(msg);
+  for (unsigned int i = 0; i < 10 && !images_received_; ++i)
+  {
+    ros::spinOnce();
+    ros::Duration(0.1).sleep();
+  }
+  EXPECT_EQ(images_received_, 1);
+
+  // then wait for 3 seconds to confirm that we don't receive any more images
+  for (unsigned int i = 0; i < 30; ++i)
+  {
+    ros::spinOnce();
+    ros::Duration(0.1).sleep();
+  }
+  EXPECT_EQ(images_received_, 1);
+
+  // then send two trigger messages very close together, but only expect one image
+  trigger_pub.publish(msg);
+  ros::spinOnce();
+  ros::Duration(0.01).sleep();
+  trigger_pub.publish(msg);
+  ros::spinOnce();
+  ros::Duration(0.01).sleep();
+  for (unsigned int i = 0; i < 10 && 1 == images_received_; ++i)
+  {
+    ros::spinOnce();
+    ros::Duration(0.1).sleep();
+  }
+  EXPECT_EQ(images_received_, 2);
+
+  // then wait for 3 seconds to confirm that we don't receive any more images
+  for (unsigned int i = 0; i < 30; ++i)
+  {
+    ros::spinOnce();
+    ros::Duration(0.1).sleep();
+  }
+  EXPECT_EQ(images_received_, 2);
```

Comment:
is there logic to filter out mgs close together?  I'm getting an error:

~~~
[gazebo_plugins.rosunit-triggered_camera/cameraSubscribeTest][FAILURE]----------
/home/osrf/code/ian/gz_ws/src/gazebo_ros_pkgs/gazebo_plugins/test/camera/triggered_camera.cpp:103
Value of: 2
Expected: images_received_
Which is: 3
~~~

but the tests are passing on jenkins..

Reply:
hm...it is kind of a race condition, but it was pretty consistent on my machine

are you consistently getting this result?

Reply:
I just ran the test 10 times and it failed once so likely a race condition.

Reply:
an alternative is to try to change the behavior so that each trigger signal leads to a new image. this would require a slight refactoring (use `int` instead of `bool` in a few places) but would be more predictable

Reply:
I think I've fixed the race condition in cb9f299. Now the plugin uses an `int` to keep track of the number of triggers received from the topic.